File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/bluetoothLE/AdvertiseStage.java
package net.ballmerlabs.uscatterbrain.network.bluetoothLE;

import androidx.annotation.Nullable;

import net.ballmerlabs.uscatterbrain.network.AdvertisePacket;

import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicReference;

public class AdvertiseStage {
    private final AtomicReference<AdvertisePacket> packet = new AtomicReference<>();
    private static final ArrayList<AdvertisePacket.Provides> provides =
            new ArrayList<AdvertisePacket.Provides>() {
        {
            add(AdvertisePacket.Provides.BLE);
            add(AdvertisePacket.Provides.WIFIP2P);
        }
    };
    private static final AdvertisePacket self = AdvertisePacket.newBuilder()
            .setProvides(provides)
            .build();

    public AdvertiseStage() {

    }


    public static AdvertisePacket getSelf() {
        return self;
    }

    public void addPacket(AdvertisePacket packet) {
        this.packet.set(packet);
    }

    @Nullable
    public AdvertisePacket getPackets() {
        return packet.get();
    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/bluetoothLE/AdvertiseStage.kt
package net.ballmerlabs.uscatterbrain.network.bluetoothLE


import net.ballmerlabs.scatterproto.Provides
import net.ballmerlabs.uscatterbrain.network.proto.AdvertisePacket
import scatterbrain.Merkle.DeclareHashesMode
import java.util.concurrent.atomic.AtomicReference

/**
 * state for bluetoothLE advertise stage
 * used in state machine implementation
 */
class AdvertiseStage : LeDeviceSession.Stage {
    private val packet = AtomicReference<AdvertisePacket>()
    var declareHashesMode = DeclareHashesMode.MERKLEPROOF
    fun addPacket(packet: AdvertisePacket) {
        declareHashesMode = packet.mode
        this.packet.set(packet)
    }

    override fun reset() {
        packet.set(null)
    }
    
    companion object {
        private val provides: ArrayList<Provides> = object : ArrayList<Provides>() {
            init {
                add(Provides.BLE)
                add(Provides.WIFIP2P)
            }
        }
        fun self(busy: Boolean = false): AdvertisePacket  {
            return AdvertisePacket.newBuilder()
                .setBusy(busy)
                .setProvides(provides)
                .build()!!
        }

    }
}